### PR TITLE
lower kusama staking limits

### PR DESCRIPTION
**Pull request by kianenigma**
This PR is part of my effort to re-configure Kusama's staking system for long term safety, and prepare it for nomination pools. 

Recall that the NPoS election process, in order to be both successful and safe, contains 4 main steps. 
1. Snapshot creation.
2. Signed submissions from staking-miner.
3. Unsigned submissions from validators, as backup. 
4. Solution verification.

Our experimentation has unraveled that, as it stands now, the bottleneck for the Kusama network (and Polkadot and Westend) is the third step, and the largest input data for which this step could be executed, in WASM, without running out of resources (weight, memory, block-length), is when there are at most around 12500 nominators in the system. Note that this number is expectedly lower than Polkadot’s 22500, because Kusama opted to allow for up to 24 nominations per nominator. Thus, we need to assume the worst case in our experimentation, namely that each nominator is indeed using all of their 24 nominations. 

More information: https://docs.google.com/document/d/1aKWbaYnzyIp7mnWlQYccrQtGLOIuVw8SD277HZuR1vw/edit?usp=sharing

### CODE REVIEW

**Comment by shawntabrizi** [comment]
> Note that this number is expectedly lower than Polkadot’s 22500, because Kusama opted to allow for up to 24 nominations per nominator. Thus, we need to assume the worst case in our experimentation, namely that each nominator is indeed using all of their 24 nominations.

I think with this in mind, it is better to reduce this number from 24 than to reduce the voter snapshot per block.

The upside of having people nominate 24 people instead of 16 people seems way smaller than the upside of including another 10,000 people into the staking system calculation logic.

**Comment by bkchr** [comment]
I'm seconding @shawntabrizi

**Comment by kianenigma** [comment]
One minor problem with going down from 24 to 16 will be that we need to give sufficient heads up for people to update their nominations, which might be tricky. If the maximum is set back to 16, all of those who are voting for more than 16 will no longer be treated as a nominator, and can also be chilled via `chill_other`.

Nonetheless, I am open to explore this path as well.

**Review by emostov — APPROVED** [review]
If we want to go down the path of 16 targets on Kusama I think that should be done as a follow up since it's more controversial and would force any nominators with more than 16 targets to change their strategy. Conversely, this current proposed change shouldn't require any intervention by active nominators and will improves safety by ensuring OCW (or any stage in the election process) won't OOM. 

I don't see any reason to punt off this fix while the community does bikesheding on the ideal number of targets.

**Comment by kianenigma** [comment]
Indeed, and looking at the numbers again, Kusama is only about halfway into filling the proposed 12_500 limit, and with the current growth rate, it would be a long time until it is reached. 

We also have PRs like https://github.com/paritytech/substrate/pull/10340 which could solve this problem in a more novel way. In the meantime though, I strongly advice we move forward with this.

**Review by shawntabrizi — APPROVED** [review]
We will use this as a stop gap until we lower the number of nominations per nominator.

File: runtime/kusama/src/lib.rs [hunk 1]
```diff
@@ -410,11 +410,7 @@ parameter_types! {
 
 	// 1 hour session, 15 minutes unsigned phase, 8 offchain executions.
 	pub OffchainRepeat: BlockNumber = UnsignedPhase::get() / 8;
-
-	/// Whilst `UseNominatorsAndUpdateBagsList` or `UseNominatorsMap` is in use, this can still be a
-	/// very large value. Once the `BagsList` is in full motion, staking might open its door to many
-	/// more nominators, and this value should instead be what is a "safe" number (e.g. 22500).
-	pub const VoterSnapshotPerBlock: u32 = 22_500;
+	pub const VoterSnapshotPerBlock: u32 = 12_500;
```
Comment (KiChjang) [inline runtime/kusama/src/lib.rs:413]: But y u remove comment?

Comment (kianenigma) [reply runtime/kusama/src/lib.rs:413]: comment was referring to the pre-bags-list world, we are passed that now :)

**Comment by kianenigma** [comment]
Will leave this open for more feedback, or pushback, but ideally it whould be merged to be part of 9190 runtime.

File: runtime/kusama/src/lib.rs (hunk 1 above)
Comment (KiChjang) [reply runtime/kusama/src/lib.rs:413]: Yeah but what I mean was, it should still have a comment explaining why we have set it to be `12_500`.

**Comment by kianenigma** [comment]
next release might branch off soon, so merging this.

**Comment by kianenigma** [comment]
bot merge

**Pull request merged**